lib/aws.js

```diff
@@ -133,6 +134,65 @@ module.exports.uploadDirectoryToS3Async = async function(s3Bucket, s3Path, local
 };
 module.exports.uploadDirectoryToS3 = util.callbackify(this.uploadDirectoryToS3Async);
 
+/**
+ * Download a file or directory from S3.
+ * 
+ * @param {string} s3Bucket - The S3 bucket name.
+ * @param {string} s3Path - The S3 source path.
+ * @param {string} localPath - The local target path.
+ * @param {object?} watcherUpdateQueue - A file watcher update queue (optional, defaults to null).
+ * @param {boolean?} forWorkspace - Whether the download is for a workspace (optional, defaults to true).
+ */
+module.exports.downloadFromS3Async = async function (s3Bucket, s3Path, localPath, watcherUpdateQueue=null, forWorkspace=true) {
+    if (localPath.slice(-1) == '/') {
+        debug(`downloadFromS3: directory localPath=${localPath}`);
+        localPath = localPath.slice(0, -1);
+        try {
+            await fs.promises.lstat(localPath);
+        } catch (err) {
+            await fs.promises.mkdir(localPath, { recursive: true });
+        }
+        if (watcherUpdateQueue != null) watcherUpdateQueue[[localPath, true]] = {action: 'skip'};
+        return;
+    } else {
+        debug(`downloadFromS3: file localPath=${localPath}`);
+        try {
+            await fs.promises.lstat(path.dirname(localPath));
+        } catch (err) {
+            await fs.promises.mkdir(path.dirname(localPath), { recursive: true });
+        }
+    }
+    if (forWorkspace) await workspaceHelper.changeWorkspaceFileOwnerAsync(localPath);
+
+    const s3 = new AWS.S3();
+    const params = {
+        Bucket: s3Bucket,
+        Key: s3Path,
+    };
+    const fileStream = fs.createWriteStream(localPath);
+    const s3Stream = s3.getObject(params).createReadStream();
+
+    return new Promise((resolve, reject) => {
+        s3Stream.on('error', function(err) {
+            debug(`downloadFromS3: missing s3://${s3Bucket}/${s3Path}`);
+            reject(err);
+        });
+        s3Stream.pipe(fileStream).on('error', function(err) {
+            debug(`downloadFromS3: connection lost`);
+            reject(err);
+        }).on('close', function() {
+            if (watcherUpdateQueue != null) watcherUpdateQueue[[localPath, false]] = {action: 'skip'};
```

Comment:
You mean by delaying it to the callback? I was thinking that would cause a race condition where chokidar might have already added it to the `update_queue` before we've added a `skip` action here. Does that make sense, or would it be fine to just add this `skip` action in the callback?